tika/server.go

```diff
@@ -0,0 +1,230 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tika
+
+import (
+	"context"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"os/exec"
+	"time"
+
+	"golang.org/x/net/context/ctxhttp"
+)
+
+// Server represents a Tika server. Create a new Server with NewServer,
+// start it with Start, and shut it down with the close function returned
+// from Start.
+// There is no need to create a Server for an already running Tika Server
+// since you can pass its URL directly to a Client.
+type Server struct {
+	jar            string
+	url            string // url is derived from port and hostname.
+	port           string
+	hostname       string
+	cancel         func()
+	startupTimeout time.Duration
+}
+
+// URL returns the URL of this Server.
+func (s *Server) URL() string {
+	return s.url
+}
+
+// An Option can be passed to NewServer to configure the Server.
+type Option func(*Server)
+
+// WithHostname returns an Option to set the host of the Server (default localhost).
+func WithHostname(h string) Option {
+	return func(s *Server) {
+		s.hostname = h
+	}
+}
+
+// WithPort returns an Option to set the port of the Server (default 9998).
+func WithPort(p string) Option {
+	return func(s *Server) {
+		s.port = p
+	}
+}
+
+// WithStartupTimeout returns an Option to set the timeout for how long to wait
+// for the Server to start (default 10s).
+func WithStartupTimeout(d time.Duration) Option {
+	return func(s *Server) {
+		s.startupTimeout = d
+	}
+}
+
+// NewServer creates a new Server.
+func NewServer(jar string, options ...Option) (*Server, error) {
+	if jar == "" {
+		return nil, fmt.Errorf("no jar file specified")
+	}
+	if _, err := os.Stat(jar); err != nil {
+		return nil, fmt.Errorf("jar file not found: %s", jar)
+	}
+	s := &Server{
+		jar:            jar,
+		port:           "9998",
+		startupTimeout: 10 * time.Second,
+		hostname:       "localhost",
+	}
+	for _, o := range options {
+		o(s)
+	}
+	urlString := "http://" + s.hostname + ":" + s.port
+	u, err := url.Parse(urlString)
+	if err != nil {
+		return nil, fmt.Errorf("invalid hostname %q or port %q: %v", s.hostname, s.port, err)
+	}
+	s.url = u.String()
+	return s, nil
+}
+
+type commander func(context.Context, string, ...string) *exec.Cmd
+
+// cmder is used to stub out *exec.Cmd for testing.
+var cmder commander = exec.CommandContext
+
+// Start starts the given server. Start will start a new Java process. The
+// caller must call cancel() to shut down the process when finished with the
+// Server. The given Context is used for the Java process, not for cancellation
+// of startup.
+func (s *Server) Start(ctx context.Context) (cancel func(), err error) {
+	ctx, cancel = context.WithCancel(ctx)
+	cmd := cmder(ctx, "java", "-jar", s.jar, "-p", s.port)
```

Comment:
It might be easier to use `Command` instead of `CommandContext` and use `cmd.Process.Kill` to stop the command from a `Stop` method on `Server`.